beetsplug/audible.py

```diff
@@ -104,11 +257,117 @@ def __init__(self):
+        stripped_titles = [strip_affixes(i.title, affixes) for i in items]
+
+        starting_numbers = [check_starts_with_number(s) for s in stripped_titles]
+        if all([s is not None for s in starting_numbers]) and is_continuous_number_series(sorted(starting_numbers)):
+            items_with_numbers = list(zip(starting_numbers, items))
+            matches = sorted(items_with_numbers, key=lambda i: i[0])
+            matches = [i[1] for i in matches]
+            return matches
+
+    def attempt_match_natsort(self, items: List[Item], album: AlbumInfo) -> Optional[List[Item]]:
+        """Use a natural sort on the input tracks to order them like a person would i.e. 10 is after 9, not 2."""
+        affixes = find_regular_affixes([c.title for c in items])
+        stripped_titles = [strip_affixes(i.title, affixes) for i in items]
+        average_title_change = calculate_average_levenshtein_difference(stripped_titles)
+        # magic number here, it's a judgement call
+        if max(average_title_change) < 4:
+            # can't assume that the tracks actually match even when there are the same number of items, since lengths
+            # can be different e.g. an even split into n parts that aren't necessarily chapter-based so just natsort
+            matches = natsorted(items, key=lambda t: t.title)
+            return matches
+
+    def attempt_match_chapter_levenshtein(self, items: List[Item], album: AlbumInfo) -> Optional[List[Item]]:
+        """For every chapter in the input album, calculate the Levenshtein difference between every item in the online
+        album and match each input track to the closest online track."""
+        # Warning, this method is rather messy, and it's easy for this to go wrong.
+        # This should be used as a last resort
+        affixes = find_regular_affixes([c.title for c in items])
+
+        all_remote_chapters: List = deepcopy(album.tracks)
+        matches = []
+        for chapter in items:
+            # TODO: need a string distance algorithm that penalises number replacements more
+            best_matches = list(
+                sorted(
+                    all_remote_chapters,
+                    key=lambda c: specialised_levenshtein(chapter.title, c.title, affixes),
+                )
+            )
+            try:
+                best_match = best_matches[0]
+            except IndexError:
+                return None
+            matches.append(best_match)
+            all_remote_chapters.remove(best_match)
+        return matches
+
+    def attempt_match_single_item(self, items: List[Item], album: AlbumInfo) -> Optional[List[Item]]:
+        """If the input album has a single item, use that; if the album also has a single item, prefer that."""
+        if len(items) == 1:
+            # Prefer a single named book from the remote source
+            if len(album.tracks) == 1 and album.tracks[0].title != "Chapter 1":
```

Comment:
Is the extra condition on chapter 1 needed? It seems safe enough to trust it if the audible source has a single track.

Would prefer removing this check if itss not needed to simplify the logic.